frontend/expr_inferrer.py

```diff
@@ -0,0 +1,372 @@
+		upper_type = infer(node.slice.upper, context)
+		step_type = infer(node.slice.step, context)
+
+		if not (lower_type.is_subtype(TInt()) and upper_type.is_subtype(TInt()) and step_type.is_subtype(TInt())):
+			raise KeyError("Slicing bounds and step should be integers.")
+		if isinstance(indexed_type, TTuple):
+			return indexed_type.get_possible_tuple_slicings()
+		else:
+			return indexed_type
+
+def infer_compare(node):
+    return TBool()
+
+def infer_name(node, context):
+    """Infer the type of a variable
+
+    Attributes:
+        node: the variable node whose type is to be inferred
+        context: The context to look in for the variable type
+    """
+    return context.get_type(node.id)
+
+def infer_generators(generators, local_context):
+    for gen in generators:
+        iter_type = infer(gen.iter, local_context)
+        if not (isinstance(iter_type, TList) or isinstance(iter_type, TSet)):
+            raise TypeError("The iterable should be only a list or a set. Found {}.", iter_type.get_name())
+        target_type = iter_type.type # Get the type of list/set elements
+
+        if not isinstance(gen.target, ast.Name):
+            raise TypeError("The iteration target should be only a variable name.")
+        local_context.set_type(gen.target.id, target_type)
+
+def infer_sequence_comprehension(node, sequence_type, context):
+    """Infer the type of a list comprehension
+
+    Attributes:
+        node: the comprehension AST node to be inferred
+        sequence_type: Either TList or TSet
+        context: The current context level
+
+    Examples:
+        - [c * 2 for c in [1,2,3]] --> [2,4,6]
+        - [c for b in [[1,2],[3,4]] for c in b] --> [1,2,3,4]
+        - [(c + 1, c * 2) for c in [1,2,3]] --> [(2,2),(3,4),(4,6)]
+
+    Limitation:
+        The iterable should always be a list or a set (not a tuple or a dict)
+        The iteration target should be always a variable name.
+    """
+    local_context = Context(parent_context=context)
+    infer_generators(node.generators, local_context)
+    return sequence_type(infer(node.elt, local_context))
+
+def infer_dict_comprehension(node, context):
+    """Infer the type of a dictionary comprehension
+
+    Attributes:
+        node: the dict comprehension AST node to be inferred
+        context: The current context level
+
+    Examples:
+        - {a:(2 * a) for a in [1,2,3]} --> {1:2, 2:4, 3:6}
+        - {a:len(a) for a in ["a","ab","abc"]}--> {"a":1, "ab":2, "abc":3}
+
+    Limitation:
+        The iterable should always be a list or a set (not a tuple or a dict)
+        The iteration target should be always a variable name.
+    """
+    local_context = Context(parent_context=context)
+    infer_generators(node.generators, local_context)
+    key_type = infer(node.key, local_context)
+    val_type = infer(node.value, local_context)
+    return TDictionary(key_type, val_type)
+
+
+def infer(node, context):
+    """Infer the type of a given AST node"""
```

Comment:
If you wanted to, you could implement this as a kind of visitor, where the infer method for ast.Num is called visit_Num, the one for ast.Str is called visit_Str etc, and then visit uses reflection to call 'visit_' + node.__class__.__name__  on self. Just an idea though.